server/emulator.go

```diff
@@ -84,49 +89,119 @@ func (m EmulatorAPIServer) CommitBlock(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	bytes, err := json.Marshal(contexts)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	_, _ = w.Write(bytes)
+
 }
 
-func (m EmulatorAPIServer) Snapshot(w http.ResponseWriter, r *http.Request) {
+func (m EmulatorAPIServer) reloadBlockchainFromSnapshot(name string, badgerStore *badger.Store, w http.ResponseWriter) {
+	blockchain, err := configureBlockchain(m.server.config, badgerStore)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	m.backend.SetEmulator(blockchain)
+	block, err := blockchain.GetLatestBlock()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	blockResponse := &BlockResponse{
+		Height:  int(block.Header.Height),
+		BlockId: block.Header.ID().String(),
+		Context: name,
+	}
+
+	bytes, err := json.Marshal(blockResponse)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	_, _ = w.Write(bytes)
+
+}
+
+func (m EmulatorAPIServer) SnapshotJump(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json")
 	vars := mux.Vars(r)
 	name := vars["name"]
 
 	switch (*m.storage).Store().(type) {
 	case *badger.Store:
 		badgerStore := (*m.storage).Store().(*badger.Store)
-		err := badgerStore.JumpToContext(name)
+		contexts, err := badgerStore.ListContexts()
 		if err != nil {
 			w.WriteHeader(http.StatusInternalServerError)
 			return
 		}
+		if !slices.Contains(contexts, name) {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
 
-		blockchain, err := configureBlockchain(m.server.config, badgerStore)
+		err = badgerStore.JumpToContext(name, false)
 		if err != nil {
 			w.WriteHeader(http.StatusInternalServerError)
 			return
 		}
 
-		m.backend.SetEmulator(blockchain)
-		block, err := blockchain.GetLatestBlock()
+		m.reloadBlockchainFromSnapshot(name, badgerStore, w)
+
+	default:
+		m.server.logger.Error("State management only available with badger storage")
+		w.WriteHeader(http.StatusInternalServerError)
+	}
+
+}
+
+func (m EmulatorAPIServer) SnapshotCreate(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+	name := r.FormValue("name")
+
+	if name == "" {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	switch (*m.storage).Store().(type) {
+	case *badger.Store:
+		badgerStore := (*m.storage).Store().(*badger.Store)
+		contexts, err := badgerStore.ListContexts()
 		if err != nil {
 			w.WriteHeader(http.StatusInternalServerError)
 			return
 		}
-
-		blockResponse := &BlockResponse{
-			Height:  int(block.Header.Height),
-			BlockId: block.Header.ID().String(),
-			Context: name,
+		if slices.Contains(contexts, name) {
+			w.WriteHeader(http.StatusConflict)
+			return
 		}
-
-		bytes, err := json.Marshal(blockResponse)
+		err = badgerStore.JumpToContext(name, true)
```

Comment:
I don't completely understand if this is necessary to jump? aren't we already at the point of snapshot?